Add tag activity reports, by user and by namespace

PurpleGuitar opened the pull request:
This change creates a new syntax plugin, 'chunkprogress'.

The plugin allows users to create two new reports, both of which examine page revision history to find status tag changes, e.g. "draft" -> "check".

Two reports are available: 
- an activity-by-user report, which displays status tag activity by user, and
- an activity-by-namespace report, which displays status tag activity for a given namespace.

PurpleGuitar committed:
Initial plugin development -- hello, world!

PurpleGuitar committed:
Simple dump of internal parms

PurpleGuitar committed:
Grab raw match of parameters

PurpleGuitar committed:
Accept and validate parameters

PurpleGuitar committed:
Accept a debug parameter

PurpleGuitar committed:
Get metadata from requested page

PurpleGuitar committed:
Testing

PurpleGuitar committed:
Lint cleanup

PurpleGuitar committed:
Updated docs to match PEAR standard

PurpleGuitar committed:
Pull revision history for a page

PurpleGuitar committed:
Create an associative array for each revision

PurpleGuitar committed:
Pull filename from revision

PurpleGuitar committed:
Grab tags from each revision of the page

PurpleGuitar committed:
Split out tags on each revision

PurpleGuitar committed:
Extract status tags from page tags for each revision

PurpleGuitar committed:
Fix some empty-array bugs

PurpleGuitar committed:
Show only pages where status has changed

PurpleGuitar committed:
Pull current page data from metadata, not last_change

PurpleGuitar committed:
Capture user who did the edit, refactor

PurpleGuitar committed:
Remove debugging code

PurpleGuitar committed:
Ignore sandbox.txt

PurpleGuitar committed:
Temporary change to main .gitignore

PurpleGuitar committed:
Remove unneeded require statment

PurpleGuitar committed:
Refactor out parameter extraction

PurpleGuitar committed:
Show blank instead of "(unknown)" for user

PurpleGuitar committed:
Doc update

PurpleGuitar committed:
Refactor out timestamp and user

PurpleGuitar committed:
Doc update

PurpleGuitar committed:
Removed superfluous code

PurpleGuitar committed:
Refactor tag extraction into a method

PurpleGuitar committed:
Refactor out getting tags and status tags

PurpleGuitar committed:
Accept report type, namespace

PurpleGuitar committed:
Find all pages under namespace

PurpleGuitar committed:
Factor out getting all pages in a namespace

PurpleGuitar committed:
Better error handling

PurpleGuitar committed:
Handle start and end dates

PurpleGuitar committed:
Get last status by user by page

PurpleGuitar committed:
Get last status by user by page

PurpleGuitar committed:
Calculate count of tags per user

PurpleGuitar committed:
Render first version of report

PurpleGuitar committed:
Corrected spacing

PurpleGuitar committed:
Capture render time

PurpleGuitar committed:
Filter by user

PurpleGuitar committed:
Cleanup

PurpleGuitar committed:
Move a few functions around

PurpleGuitar committed:
Factor out namespace validation

PurpleGuitar committed:
Factor out start date validation

PurpleGuitar committed:
Refactor out end date validation

PurpleGuitar committed:
Refactor out activity-by-user report into its own file

PurpleGuitar committed:
Got rid of some unnecessary variables

PurpleGuitar committed:
Create shell of activity-by-namespace report

PurpleGuitar committed:
Round off elapsed time

PurpleGuitar committed:
Grab pages and split by sub-namespace (e.g. book)

PurpleGuitar committed:
Look up revisions for pages

PurpleGuitar committed:
Fix bug getting statuses

PurpleGuitar committed:
Remember current revision of the page

PurpleGuitar committed:
Create activity by namespace report!

PurpleGuitar committed:
Merge from development

PurpleGuitar committed:
Add total line to activity-by-namespace report

PurpleGuitar committed:
Add total row to activity-by-user report

PurpleGuitar committed:
Merge branch 'development' into cro-dev

PurpleGuitar committed:
Remove unneeded .gitignore files

PurpleGuitar committed:
Restore dev .gitignore files, getting ready for merge into development

phillip-hopper commented:
@PurpleGuitar, could you add some unit tests to show that this plugin does what it should?

PurpleGuitar committed:
Add documentation for running reports

Includes a markdown document and some supporting images in the doc/ directory.
The report can be easily rendered to HTML or PDF using your favorite markdown
tool. (I use pandoc.)

PurpleGuitar committed:
Added an initial unit test.

It's a copy of the general.test.php from door43obs.  Confirmed to execute and
pass without errors.

PurpleGuitar committed:
Add test data and second test

Second test isn't running correctly yet.

PurpleGuitar committed:
Chunk progress plugin runs, but still missing data

PurpleGuitar committed:
Got test working by putting data in the right place

PurpleGuitar committed:
Verify namespace report contents

PurpleGuitar committed:
Add unit test for activity-by-user report

PurpleGuitar commented:
OK, I added three tests, two of which are regression tests verifying the reports' output.

alerque commented:
@PurpleGuitar You might also add the Wycliff email address that you've been commiting to git with to your Github account so that it associates these commits with you. The link will make it a touch easier to follow conversations across issues, commits, PR's, etc.
